esp-metadata/src/cfg/gpio.rs

```diff
@@ -0,0 +1,358 @@
+//! This module contains configuration used in [device.gpio], as well as
+//! functions that generate code for esp-hal.
+
+use std::str::FromStr;
+
+use proc_macro2::TokenStream;
+use quote::format_ident;
+
+use crate::number;
+
+/// Pin capabilities. Some of these will cause a trait to be implemented for the
+/// given pin singleton. `UsbDevice` currently only changes what happens on GPIO
+/// driver initialization.
+#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
+#[serde(rename_all = "snake_case")]
+pub(crate) enum PinCapability {
+    Input,
+    Output,
+    Analog,
+    Rtc,
+    Touch,
+    UsbDm,
+    UsbDp,
+    // Pin has USB pullup according to the IO MUX Function list
+    UsbDevice,
+}
+
+/// A row in the IO MUX Pad List table. Used in
+/// [device.gpio.pins[X].alternate_functions]. The GPIO function is not written
+/// here as that is common to all pins. The values are signal names listed in
+/// [device.gpio.input_signals] or [device.gpio.output_signals]. `None` means
+/// the pin does not provide the given alternate function.
+#[derive(Debug, Default, Clone, serde::Deserialize, serde::Serialize)]
+pub(crate) struct AfMap {
+    #[serde(rename = "0")]
+    af0: Option<String>,
+    #[serde(rename = "1")]
+    af1: Option<String>,
+    #[serde(rename = "2")]
+    af2: Option<String>,
+    #[serde(rename = "3")]
+    af3: Option<String>,
+    #[serde(rename = "4")]
+    af4: Option<String>,
+    #[serde(rename = "5")]
+    af5: Option<String>,
+}
+
+impl AfMap {
+    pub fn get(&self, af: usize) -> Option<&str> {
```

Comment:
I'm curious if it makes sense to add a docstring here (and in other public API, for example pub fields in structs)?

Reply:
Thanks for adding them!